### ISSUE TITLE
[BUG] Model weights are in low precision if bf16 or fp16 is set to True

### ISSUE THREAD

**Issue by Monohydroxides** [microsoft/DeepSpeed#6640]
**Describe the bug**
When Deepspeed uses bf16 or fp16 for training, model weights will be converted to 16-bit precision instead of keeping in 32-bit precision:
https://github.com/microsoft/DeepSpeed/issues/4772#issuecomment-1873695091
https://github.com/Lightning-AI/pytorch-lightning/pull/18113#issuecomment-1642304712

Code: https://github.com/microsoft/DeepSpeed/blob/6eefc3d0ead2c6360eec248daab0cae66a737ea0/deepspeed/runtime/engine.py#L1142-L1151

In some cases, 16-bit precision may lead to significant performance issues:
https://github.com/microsoft/DeepSpeed/issues/5307#issue-2204133226

When we set bf16 or fp16 to True, is there a way to keep the model weights in 32-bit precision to mitigate potential performance issues? This should be considered a bug rather than a design.

**Expected behavior**
When bf16 or fp16 is set to True, a solution should still be provided to keep the model parameters in 32-bit precision.

**Comment by jomayeri** [comment]
The master weights in the optimizer are still in full precision fp32.

**Comment by Monohydroxides** [comment]
> The master weights in the optimizer are still in full precision fp32.

Some operations that require high precision (softmax, layernorm) are also in 16-bit precision during both forward and backward propagation, and this is exactly what confuses me.

**Comment by Monohydroxides** [comment]
> The master weights in the optimizer are still in full precision fp32.

Yes, but my concern is whether it is appropriate to use only half-precision during forward and backward propagation. There's no doubt that we should use FP32 when parameter updates, but is it possible that DeepSpeed provides some mechanism to retain specified weights in FP32 for computation during forward and backward propagation?

**Comment by jomayeri** [comment]
You are able to train in fp32 parameters if you don't select bf16: true, or fp16: true. What is not available is some parameters in fp32 and some in fp16.

**Comment by Monohydroxides** [comment]
Got it. Close issue.
